lxd/db_storage_pools.go

```diff
@@ -0,0 +1,520 @@
+	tmp := storagePoolDriversCacheVal.Load()
+	if tmp != nil {
+		typedTmp := tmp.([]string)
+		if !shared.StringInSlice(poolDriver, typedTmp) {
+			typedTmp = append(typedTmp, poolDriver)
+		}
+		storagePoolDriversCacheVal.Store(typedTmp)
+	}
+	storagePoolDriversCacheMu.Unlock()
+
+	return id, nil
+}
+
+// Add new storage pool config.
+func dbStoragePoolConfigAdd(tx *sql.Tx, poolID int64, poolConfig map[string]string) error {
+	str := "INSERT INTO storage_pools_config (storage_pool_id, key, value) VALUES(?, ?, ?)"
+	stmt, err := tx.Prepare(str)
+	defer stmt.Close()
+
+	for k, v := range poolConfig {
+		if v == "" {
+			continue
+		}
+
+		_, err = stmt.Exec(poolID, k, v)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Update storage pool.
+func dbStoragePoolUpdate(db *sql.DB, poolName string, poolConfig map[string]string) error {
+	poolID, _, err := dbStoragePoolGet(db, poolName)
+	if err != nil {
+		return err
+	}
+
+	tx, err := dbBegin(db)
+	if err != nil {
+		return err
+	}
+
+	err = dbStoragePoolConfigClear(tx, poolID)
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	err = dbStoragePoolConfigAdd(tx, poolID, poolConfig)
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	return txCommit(tx)
+}
+
+// Delete storage pool config.
+func dbStoragePoolConfigClear(tx *sql.Tx, poolID int64) error {
+	_, err := tx.Exec("DELETE FROM storage_pools_config WHERE storage_pool_id=?", poolID)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Delete storage pool.
+func dbStoragePoolDelete(db *sql.DB, poolName string) error {
+	poolID, pool, err := dbStoragePoolGet(db, poolName)
+	if err != nil {
+		return err
+	}
+
+	tx, err := dbBegin(db)
+	if err != nil {
+		return err
+	}
+
+	_, err = tx.Exec("DELETE FROM storage_pools WHERE id=?", poolID)
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	err = dbStoragePoolConfigClear(tx, poolID)
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	// COMMENT(brauner): Update the storage drivers cache in api_1.0.go.
+	storagePoolDriversCacheMu.Lock()
+	tmp := storagePoolDriversCacheVal.Load()
+	if tmp != nil {
+		typedTmp := tmp.([]string)
+		for i := 0; i < len(typedTmp); i++ {
+			if typedTmp[i] == pool.PoolDriver {
+				typedTmp[i] = typedTmp[len(typedTmp)-1]
+				typedTmp[len(typedTmp)-1] = ""
+				typedTmp = typedTmp[:len(typedTmp)-1]
+				break
+			}
+		}
+		storagePoolDriversCacheVal.Store(typedTmp)
+	}
+	storagePoolDriversCacheMu.Unlock()
+
+	return txCommit(tx)
+}
+
+// Get the names of all storage volumes attached to a given storage pool.
+func dbStoragePoolVolumesGetNames(db *sql.DB, poolID int64) ([]string, error) {
+	var volumeName string
+	query := "SELECT DISTINCT name FROM storage_volumes WHERE storage_pool_id=?"
+	inargs := []interface{}{poolID}
+	outargs := []interface{}{volumeName}
+
+	result, err := dbQueryScan(db, query, inargs, outargs)
+	if err != nil {
+		return []string{}, err
+	}
+
+	if len(result) == 0 {
+		return []string{}, NoSuchObjectError
```

Comment:
@tych0 there seems to be precedent for that check in the LXD codebase. Based on other such checks, it looks like QueryRowScan will get you the error when you don't get any result but QueryScan will not, it will instead just return an empty result list that you have to check for.